File: java/solitaire/deck/src/main/java/org/silnith/deck/Suit.java
package org.silnith.deck;

/**
 * A suit for a playing card.
 * 
 * @author <a href="mailto:silnith@gmail.com">Kent Rosenkoetter</a>
 */
public enum Suit {
    /**
     * &#x2663;
     */
    CLUB(Color.BLACK),
    /**
     * &#x2666;
     */
    DIAMOND(Color.RED),
    /**
     * &#x2665;
     */
    HEART(Color.RED),
    /**
     * &#x2660;
     */
    SPADE(Color.BLACK);
    
    /**
     * A color for a playing card&#8217;s suit.
     * 
     * @author <a href="mailto:silnith@gmail.com">Kent Rosenkoetter</a>
     */
    public enum Color {
        /**
         * &#x2663; &amp; &#x2660;
         * 
         * @see Suit#CLUB
         * @see Suit#SPADE
         */
        BLACK,
        /**
         * &#x2666; &amp; &#x2665;
         * 
         * @see Suit#DIAMOND
         * @see Suit#HEART
         */
        RED
    }
    
    private final Color color;
    
    private Suit(final Color color) {
    	assert color != null;
        this.color = color;
    }
    
    /**
     * Returns the color of a particular suit.
     * 
     * @return the color of the suit
     */
    public Color getColor() {
        return color;
    }
    
    /**
     * Returns the Unicode symbol for the card suit.
     * <table>
     * <caption>The Unicode symbols for each card suit.</caption>
     * <thead>
     * <tr>
     * <th>Suit</th>
     * <th>Symbol</th>
     * </tr>
     * </thead> <tbody>
     * <tr>
     * <td>{@link #CLUB}</td>
     * <td>&#x2663;</td>
     * </tr>
     * <tr>
     * <td>{@link #DIAMOND}</td>
     * <td>&#x2666;</td>
     * </tr>
     * <tr>
     * <td>{@link #HEART}</td>
     * <td>&#x2665;</td>
     * </tr>
     * <tr>
     * <td>{@link #SPADE}</td>
     * <td>&#x2660;</td>
     * </tr>
     * </tbody>
     * </table>
     * 
     * @return a single Unicode glyph
     */
    public String toSymbol() {
        switch (this) {
        case CLUB:
            return "\u2663";
        case DIAMOND:
            return "\u2666";
        case HEART:
            return "\u2665";
        case SPADE:
            return "\u2660";
        default:
        	assert false : "Unknown suit: " + this;
            throw new NullPointerException();
        }
    }
    
    @Override
    public String toString() {
        switch (this) {
        case CLUB:
            return "Club";
        case DIAMOND:
            return "Diamond";
        case HEART:
            return "Heart";
        case SPADE:
            return "Spade";
        default:
        	assert false : "Unknown suit: " + this;
            return super.toString();
        }
    }
    
}

File: csharp/Solitaire/Deck/Suit.cs
using System;

namespace Silnith.Game.Deck
{
    /// <summary>
    /// A suit for a playing card.
    /// </summary>
    public enum Suit
    {
        /// <summary>
        /// &#x2663;
        /// </summary>
        Club,
        /// <summary>
        /// &#x2666;
        /// </summary>
        Diamond,
        /// <summary>
        /// &#x2665;
        /// </summary>
        Heart,
        /// <summary>
        /// &#x2660;
        /// </summary>
        Spade
    }

    /// <summary>
    /// Extension methods for <see cref="Suit"/>.
    /// </summary>
    public static class SuitExtensions
    {
        /// <summary>
        /// Returns the color of a particular suit.
        /// </summary>
        /// <param name="suit">The suit.</param>
        /// <returns>The color of the suit.</returns>
        public static Color GetColor(this Suit suit)
        {
            return suit switch
            {
                Suit.Club => Color.Black,
                Suit.Diamond => Color.Red,
                Suit.Heart => Color.Red,
                Suit.Spade => Color.Black,
                _ => throw new ArgumentException("Unknown Suit", nameof(suit)),
            };
        }

        /// <summary>
        /// Returns the Unicode symbol for the card suit.
        /// </summary>
        /// <remarks>
        /// <list type="table">
        ///   <listheader>
        ///     <term>Suit</term>
        ///     <term>Symbol</term>
        ///   </listheader>
        ///   <item>
        ///     <term><see cref="Suit.Club"/></term>
        ///     <term>&#x2663;</term>
        ///   </item>
        ///   <item>
        ///     <term><see cref="Suit.Diamond"/></term>
        ///     <term>&#x2666;</term>
        ///   </item>
        ///   <item>
        ///     <term><see cref="Suit.Heart"/></term>
        ///     <term>&#x2665;</term>
        ///   </item>
        ///   <item>
        ///     <term><see cref="Suit.Spade"/></term>
        ///     <term>&#x2660;</term>
        ///   </item>
        /// </list>
        /// </remarks>
        /// <param name="suit">The suit.</param>
        /// <returns>A single Unicode glyph.</returns>
        public static string ToSymbol(this Suit suit)
        {
            return suit switch
            {
                Suit.Club => "\u2663",
                Suit.Diamond => "\u2666",
                Suit.Heart => "\u2665",
                Suit.Spade => "\u2660",
                _ => throw new ArgumentException("Unknown Suit", nameof(suit)),
            };
        }
    }
}
